### PR TITLE
Add Nutrient Web SDK demo: full host-app shell around the viewer

### PR DESCRIPTION
**Pull request by MahmoudElsayad**
## What's special about this demo

Most existing Web SDK examples in this repo show **one feature**: a single custom button, a single custom tooltip, a single replacement slot. They live *inside* the SDK's chrome.

This demo does the opposite — it **hides the SDK's chrome and rebuilds a complete document workspace around the viewer**, in the style of a host application like Dropbox's PDF UI. The viewer is just a content surface; the toolbar, file explorer, signing flow, form-creator panel, and drawing/text tools are all host-app code talking to the SDK through its instance API.

That's the angle this example covers that no other example in the repo does: it stress-tests the **app/viewer boundary** end-to-end and shows that customer-owned UI *around* the SDK is a viable path, not just per-component customization.

### The six integration patterns it demonstrates

1. **External floating ink toolbar.** Activates \`NutrientViewer.InteractionMode.INK\`, sets app-owned annotation presets through \`instance.setAnnotationPresets\` / \`instance.setCurrentAnnotationPreset\`, and drives undo/redo through \`instance.history.undo()\`. The host owns the toolbar; the SDK just renders the strokes.
2. **Hide and replace the main toolbar.** \`ui.tools.main\` is set to a render-nothing slot and the demo's own React toolbar takes over page navigation, rotate/add/delete/move, zoom, search, and download — all driven through \`instance.setViewState\` and \`instance.applyOperations\`.
3. **App-owned signature capture modal.** Four-tab dialog (Draw / Type / Upload / Saved) running on a custom canvas. Output goes through \`instance.createAttachment(blob)\` → \`ImageAnnotation({ isSignature: true, imageAttachmentId, contentType: 'image/png' })\` → \`instance.create()\`. \`signatures.create\` and \`signatures.list\` slots are bridged so the SDK's own dialog never opens.
4. **Drag-and-drop from an external panel into the viewer.** Drop listeners are attached to \`instance.contentDocument\` (the SDK runs in an iframe), then drop coordinates are converted with \`instance.transformContentClientToPageSpace(new NutrientViewer.Geometry.Point(...), pageIndex)\` and linked widget + form-field objects are created.
5. **Custom Form Creator property editor.** \`ui.formCreator.propertyEditor\` is replaced with a host-rendered panel that walks the form-field → annotation pair to rename and delete fields through \`instance.update\` / \`instance.delete\`.
6. **Custom annotation tooltip actions.** \`annotationTooltipCallback\` injects Delete-field and Edit-field buttons on top of the SDK's built-in tooltip. The Edit button flips into Form Creator mode and a separate listener restores the previous interaction mode when the selection clears.

### The API surfaces it exercises across the boundary

Slot replacement: \`ui.tools.main\`, \`ui.tools.contextual\`, \`ui.signatures.create\`, \`ui.signatures.list\`, \`ui.formCreator.propertyEditor\`, \`annotationTooltipCallback\`.

Instance methods: \`setUI\`, \`setViewState\`, \`applyOperations\`, \`history.undo\`/\`redo\`, \`setAnnotationPresets\`, \`setCurrentAnnotationPreset\`, \`contentDocument\`, \`transformContentClientToPageSpace\`, \`createAttachment\`, \`create\`, \`update\`, \`delete\`, \`getSelectedAnnotations\`, \`setSelectedAnnotations\`, \`getFormFields\`, \`getAnnotations\`, \`exportPDF\`, \`InteractionMode.*\`.

Annotation/form model: \`Annotations.WidgetAnnotation\`, \`Annotations.ImageAnnotation\` (with \`isSignature: true\`), \`FormFields.TextFormField\`, \`FormFields.CheckBoxFormField\`, \`FormOption\`, \`Geometry.Rect\`, \`Geometry.Point\`.

This breadth across one app is the point — it shows that all the pieces a customer needs to build a full host-app shell are present and that they compose.

### Why this matters

This example was originally built as a validation example for the headless / API-first strategy: prove that the SDK is usable as an embeddable viewer surface inside a customer-owned UI, not only as a closed UI box that can be poked at the edges. As a side benefit, the demo surfaced an API ergonomics gap in slot callbacks (which then drove a follow-up SDK change) — exactly the kind of feedback this category of example is meant to produce.

## What's rebuilt vs. consumed

- **Rebuilt from scratch:** main toolbar, signature dialog, Form Creator property editor, floating ink/text toolbars, file explorer, signers panel.
- **Augmented:** annotation tooltip (custom Delete/Edit items on top of the SDK's tooltip).
- **Consumed as-is:** page rendering, thumbnails sidebar, search UI, rich-text editor, annotation/form-field data model, history stack, attachment storage.

## Notes

- The SDK is loaded from the CDN via a \`<script>\` tag (\`pspdfkit-web@1.15.0\`), so there's **no \`pspdfkit-lib\` copy step** and no \`pspdfkit\` npm dependency. \`baseUrl\` is omitted — the SDK auto-detects it from the script's origin.
- The grouped UI slots used here (\`tools.main\`, \`signatures.create\`, \`formCreator.propertyEditor\`) require Web SDK **1.14+**. Older CDN versions reject the config with \`'tools' is not a valid UI element\`.
- A shared Baseline UI dark theme is passed straight into the SDK iframe via \`NutrientViewer.load({ theme })\`, so the demo shell and the viewer chrome render with one set of color tokens — no CSS overrides, no runtime sync.
- \`biome.json\` is updated to opt the new example out of repo-wide formatting, matching the convention every other example under \`web/\` follows (\`!web/viewer/multi-tab\`, \`!web/signing/signing-demo-complete\`, etc.).

## Test plan

- [ ] \`cd web/viewer/web-sdk-demo && npm install && npm run dev\`
- [ ] Open <http://localhost:5173> — bundled \`example.pdf\` renders with the custom dark theme
- [ ] Drag a PDF into the sidebar — uploads, switches active doc
- [ ] Toolbar: page nav, rotate, add/delete/move page, zoom, download
- [ ] Sign popover → Get signatures → drag a Signature pill onto the page → signing modal opens at the drop point
- [ ] Drawing mode → floating ink toolbar appears; colour + width work; Undo + Delete drawings work
- [ ] Text mode → floating text toolbar; colour, size, bold/italic, alignment apply to a selected text annotation
- [ ] Edit field (annotation tooltip) → switches to Form Creator, custom property editor renders; renaming via the input updates both the form-field and the widget; clicking off restores the previous interaction mode
- [ ] \`npm run build\` succeeds

### CODE REVIEW

**Comment by MahmoudElsayad** [comment]
@sashamilenkovic @sc0 This is ready for you.

File: web/viewer/web-sdk-demo/src/App.tsx [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+        setSigningModal(null)
+        return
+      }
+      try {
+        const sdk = window.NutrientViewer!
+        const ImageAnnotation = sdk.Annotations?.ImageAnnotation ?? sdk.ImageAnnotation
+        const Rect = sdk.Geometry?.Rect ?? sdk.Rect
+        if (!ImageAnnotation || !Rect) {
+          console.warn('SDK ImageAnnotation/Rect not found — signature not inserted.')
+          return
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/App.tsx:302]: This `return` exits `insertSignatureIntoSelected` without `setSigningModal(null)`, so when the SDK's `ImageAnnotation`/`Rect` constructors aren't found the signing modal stays open with no feedback — and re-clicking Insert keeps re-failing. Every other exit (`!instance` at 291, success/catch at 326) clears it. Move the close into a `finally`.

File: web/viewer/web-sdk-demo/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+        }
+
+        const res = await fetch(dataUrl)
+        const blob = await res.blob()
+        const contentType = blob.type || getDataUrlContentType(dataUrl) || 'image/png'
+        const attachmentId = instance.createAttachment
+          ? await instance.createAttachment(blob)
+          : null
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/App.tsx]: When `createAttachment` is absent, `attachmentId` is `null` but the code still builds `new ImageAnnotation({ imageAttachmentId: null, … })` and `create`s it — producing a blank/throwing signature rather than failing fast. Bail with a visible message when `attachmentId` is null instead of flowing it downstream.

File: web/viewer/web-sdk-demo/src/App.tsx [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+  )
+
+  useEffect(() => {
+    return () => {
+      files.forEach((f) => {
+        if (!f.isBuiltin) URL.revokeObjectURL(f.url)
+      })
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/App.tsx:96]: The empty dep array makes this cleanup close over the first-render `files` (`[BUILTIN_FILE]`), so blob URLs created for later-uploaded files are never revoked on unmount — the `exhaustive-deps` disable is hiding exactly that stale closure. Track files in a ref and revoke `ref.current` here, or drop the effect since it currently does nothing useful.

File: web/viewer/web-sdk-demo/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+                values: list([]),
+                ...(FormOption
+                  ? { options: list([new FormOption({ label: 'Yes', value: 'Yes' })]) }
+                  : {}),
+              })
+            : TextFormField && new TextFormField({ name: formFieldName, annotationIds })
+
+        if (!formField) {
+          console.warn('SDK FormField constructor not found for field type:', type)
+          await instance.create(widget)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/App.tsx]: This fallback creates a widget carrying a `formFieldName` but **no backing form field**. That orphan is what later breaks `renameField` (it renames the widget to a name no field has) and leaves a non-functioning field on the page. Consider not creating the widget at all when the `FormField` constructor is missing, or surfacing the failure.

File: web/viewer/web-sdk-demo/src/Toolbar.tsx [hunk 5]
```diff
@@ -0,0 +1,730 @@
...
+    if (!instance) {
+      setPageIndex(0)
+      setPageCount(0)
+      setZoom(1)
+      setPageInput('1')
+      return
+    }
+
+    setPageIndex(instance.viewState.currentPageIndex)
+    setPageCount(instance.totalPageCount)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/Toolbar.tsx]: `pageCount` is set once here and only `viewState.change`/`zoom.change` are subscribed — nothing re-reads `instance.totalPageCount` after `applyOperations` add/remove/move. So the "of N" readout, the move-form validation, and the `pageCount <= 1` delete-guard all go stale until reload (add pages → count stays old; delete toward 1 → guard checks a stale count). Re-read `totalPageCount` after each operation.

File: web/viewer/web-sdk-demo/src/Toolbar.tsx [hunk 6]
```diff
@@ -0,0 +1,730 @@
...
+
+  const insertImage = async () => {
+    if (!instance) return
+    const input = document.createElement('input')
+    input.type = 'file'
+    input.accept = 'image/*'
+    input.onchange = () => {
+      // Triggering image insertion requires deeper SDK plumbing (attachments +
+      // image annotations). For this demo, we just enter PAN mode and log.
+      console.info('TODO: wire insertImage to instance.create()', input.files)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/Toolbar.tsx]: This handler opens a file picker and then only `console.info`s a TODO — the prominent "Insert image" button (line 312) is a dead no-op in a demo whose whole point is showing capabilities. The `createAttachment` → `ImageAnnotation` pattern already exists in `App.tsx insertSignatureIntoSelected`; either wire it or remove the button. Also the comment claims it "enter[s] PAN mode" — it doesn't.

File: web/viewer/web-sdk-demo/src/Toolbar.tsx [hunk 7]
```diff
@@ -0,0 +1,730 @@
...
+    if (!instance) return
+    const fromIndex = fromOneBased - 1
+    if (fromIndex < 0 || fromIndex >= pageCount) return
+
+    // The user's mental model: "After page N" means the moved page should end up
+    // at slot N+1 in the resulting document. Compute that target final index and
+    // translate it into the SDK's afterPageIndex/beforePageIndex (which reference
+    // the *original* page positions).
+    const targetIndex =
+      afterOneBased <= 0 ? 0 : Math.min(afterOneBased, pageCount - 1)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/Toolbar.tsx]: `targetIndex` mixes a 1-based "after page N" value with a 0-based final index, and clamping to `pageCount - 1` means "move to the very end" is unreachable. Can you confirm "move page 1 after page 3" (5-page doc) and "move to last" land where expected? I couldn't run it to verify the backward-move `targetIndex - 1` branch.

File: web/viewer/web-sdk-demo/src/Toolbar.tsx [hunk 8]
```diff
@@ -0,0 +1,730 @@
...
+    const buffer = await instance.exportPDF()
+    const blob = new Blob([buffer], { type: 'application/pdf' })
+    const url = URL.createObjectURL(blob)
+    const a = document.createElement('a')
+    a.href = url
+    a.download = fileName ?? 'document.pdf'
+    document.body.appendChild(a)
+    a.click()
+    document.body.removeChild(a)
+    URL.revokeObjectURL(url)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/Toolbar.tsx]: Revoking the object URL synchronously right after `a.click()` can cancel the download in some browsers before navigation commits.

```suggestion
    setTimeout(() => URL.revokeObjectURL(url), 0)
```

**Review by ritz078 — COMMENTED** [review]
Automated review pass — inline suggestions below, none blocking.

File: web/viewer/web-sdk-demo/src/Toolbar.tsx [hunk 9]
```diff
@@ -0,0 +1,730 @@
...
+    const target = Math.max(1, Math.min(pageCount, oneBasedPage)) - 1
+    apply((vs) => vs.set('currentPageIndex', target))
+  }
+
+  const zoomIn = () => apply((vs) => vs.zoomIn())
+  const zoomOut = () => apply((vs) => vs.zoomOut())
+  const fitWidth = () =>
+    apply((vs) => vs.set('zoom', window.NutrientViewer!.ZoomMode.FIT_TO_WIDTH))
+
+  const rotatePages = async (rotateBy: 90 | -90) => {
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/Toolbar.tsx]: These mutation handlers (`rotatePages`, `deletePage`, `addPage`, `movePages`, `download`) are `async` with no `.catch`/try-catch and are wired directly to `onClick`. A rejected `applyOperations`/`exportPDF` becomes an unhandled rejection and the user sees a dead button with nothing surfaced. Wrap with a visible error (or at least log).

File: web/viewer/web-sdk-demo/src/text/TextToolbar.tsx [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+
+  // `horizontalAlign` isn't in the editor namespace — it's a per-annotation
+  // property (not per-character), so we update the model directly.
+  const applyAlign = useCallback(
+    (next: Align) => {
+      setAlign(next)
+      if (!instance) return
+      const selected = getFirstSelectedAnnotation(instance.getSelectedAnnotations?.())
+      if (isTextAnnotation(selected) && selected.horizontalAlign !== next) {
+        void instance.update(selected.set('horizontalAlign', next))
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/text/TextToolbar.tsx]: `void` discards the promise, and `setAlign(next)` already flipped the local state before this call — so on a rejected `update` the toolbar shows the new alignment while the document didn't change (silent UI↔document divergence + unhandled rejection). Await and reconcile state on failure.

File: web/viewer/web-sdk-demo/src/ink/InkToolbar.tsx [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+      //    annotation is effectively committed; mutating it would re-color
+      //    drawings the user has already made. The next stroke will start a
+      //    fresh annotation that picks up the updated preset. Mirrors the
+      //    SDK's own `InkAnnotationToolbarComponent` behaviour.
+      const selected = getFirstSelectedAnnotation(instance.getSelectedAnnotations?.())
+      if (isPristineInkAnnotation(selected)) {
+        const updated = selected
+          .set('strokeColor', newStrokeColor)
+          .set('lineWidth', lineWidth)
+        void instance.update(updated)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/ink/InkToolbar.tsx]: Same fire-and-forget as the text toolbar: `void instance.update(updated)` after the local state already changed leaves the UI showing a value the document may not have. Await and handle rejection.

File: web/viewer/web-sdk-demo/src/ink/InkToolbar.tsx [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+  }, [instance])
+
+  const deleteAllInkOnPage = useCallback(async () => {
+    if (!instance) return
+    try {
+      const pageIndex = instance.viewState.currentPageIndex
+      const list = await instance.getAnnotations(pageIndex)
+      const inkIds = list
+        .toArray()
+        .filter((a) => a.constructor?.name === 'InkAnnotation')
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/ink/InkToolbar.tsx]: Filtering by `a.constructor?.name === 'InkAnnotation'` relies on class names surviving the minified CDN bundle — if they're mangled, "delete drawings on page" silently matches nothing. `instanceof window.NutrientViewer.Annotations.InkAnnotation` is robust regardless. Same pattern in `text/TextToolbar.tsx:206`. Worth confirming against the 1.15.0 CDN build.

File: web/viewer/web-sdk-demo/src/form-creator/index.ts [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+  instance: SDKInstance,
+  annotation: WidgetLikeAnnotation,
+  nextName: string,
+) {
+  try {
+    const updates: unknown[] = []
+    const formField = await getFormFieldForAnnotation(instance, annotation)
+
+    if (formField?.set) updates.push(formField.set('name', nextName))
+    if (annotation.set) updates.push(annotation.set('formFieldName', nextName))
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/form-creator/index.ts]: This rename isn't atomic across the widget↔form-field pair: if the form field can't be resolved (e.g. a widget created via the `placeFieldAt` fallback with no backing field), only `annotation.set('formFieldName', …)` runs and the widget desyncs from any field, with no rollback. `deleteField` (line 236) has the sibling issue — it falls back to `annotation.id`, deleting the widget and orphaning the FormField.

File: web/viewer/web-sdk-demo/src/types/global.d.ts [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+  setViewState: (
+    update: SDKViewState | ((vs: SDKViewState) => SDKViewState),
+  ) => void
+  exportPDF: () => Promise<ArrayBuffer>
+  applyOperations: (ops: Array<Record<string, unknown>>) => Promise<unknown>
+  create: (annotationOrList: unknown) => Promise<unknown>
+  delete: (idOrList: unknown) => Promise<unknown>
+  update: (objectOrList: unknown) => Promise<unknown>
+  createAttachment: (blob: Blob) => Promise<string>
+  getFormFields?: () => Promise<{
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/types/global.d.ts]: `getFormFields` is an always-present core SDK method but is typed optional, so `instance.getFormFields?.()` short-circuits to `null` → `getFormFieldForAnnotation` returns `[]` → `deleteField`/`renameField` silently take the orphan/desync paths. Making it required turns a genuinely-missing API into a caught throw instead of document corruption.

File: web/viewer/web-sdk-demo/tsconfig.app.json [hunk 15]
```diff
@@ -0,0 +1,21 @@
+{
+  "compilerOptions": {
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/tsconfig.app.json:2]: The build script is `tsc -b && vite build`, and the root tsconfig uses project references, but neither leaf config sets `"composite": true`. TS build mode errors **TS6306** ("Referenced project must have setting composite: true") when references lack it. I couldn't run `tsc` here — please confirm `npm run build` passes; if it errors TS6306, add `"composite": true` to both leaf `compilerOptions` (allowed alongside `noEmit` on TS 5.7). Same for `tsconfig.node.json`.

File: web/viewer/web-sdk-demo/src/text/TextToolbar.tsx [hunk 16]
```diff
@@ -0,0 +1,522 @@
...
+      console.warn('Could not update text preset', err)
+    }
+  }, [instance, color, fontSizeId, isBold, isItalic, align])
+
+  // Per-property handlers: update local state AND push to the SDK editor
+  // namespace so the change reaches the rich-text editor's current text
+  // selection (Slate transforms) when in EDITING mode, or the annotation
+  // model when only SELECTED. Each handler is the user's explicit intent —
+  // we only push the property they actually changed.
+  const applyColor = useCallback(
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/text/TextToolbar.tsx]: `applyColor`/`applyFontSize`/`applyBold`/`applyItalic` repeat an identical guard preamble (`instance` → `instance.annotations?.text` → `hasSingleTextAnnotationSelected` → `safeCall`); only the final call differs. Extracting a `withTextApi(fn)` helper collapses the four bodies with no behavior change.

File: web/viewer/web-sdk-demo/src/NutrientViewer.tsx [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+        annotationTooltipCallback: (annotation: unknown) =>
+          createAnnotationTooltipItems(loadedInstance, annotation as { formFieldName?: string }),
+        ui: initialUIConfigRef.current,
+        initialViewState: new sdk.ViewState({
+          sidebarMode: sdk.SidebarMode.THUMBNAILS,
+          sidebarWidth: 180,
+        }),
+      })
+      .then((instance) => {
+        if (cancelled) {
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/NutrientViewer.tsx:108]: This `sdk.unload(container)` (and the one in the cleanup `return`) unloads the shared container node, not the instance *this* run created. Under React 18 dev StrictMode the effect runs mount→cleanup→mount on the same container, so when this `load()` resolves late it can tear down the instance the second mount just created — blank viewer in `npm run dev`. (Production file-switches remount via `key`, so they're unaffected.) Scope the unload to the instance this effect loaded — resolve into a local and unload that — rather than `container`.

File: web/viewer/web-sdk-demo/src/lib/sdk.ts [hunk 18]
```diff
@@ -0,0 +1,29 @@
+import type { SDKInstance } from '../types/global'
+
+export function isFormCreatorMode(instance: SDKInstance): boolean {
+  const formCreator = window.NutrientViewer?.InteractionMode?.FORM_CREATOR
+  return Boolean(formCreator && instance.viewState.interactionMode === formCreator)
+}
+
+/**
+ * `instance.setSelectedAnnotations` expects either an Immutable.List of ids or
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/lib/sdk.ts]: This isn't accurate: `setSelectedAnnotations` always takes a `NutrientViewer.Immutable.List` of annotations/ids (`AnnotationSelectionMixin`) — there's no plain-array variant “depending on SDK build.” The runtime array fallback is defensive, not a documented contract. Worth correcting since this helper gets copied verbatim.

File: web/viewer/web-sdk-demo/src/lib/selection.ts [hunk 19]
```diff
@@ -0,0 +1,42 @@
+/**
+ * `instance.getSelectedAnnotations()` doesn't have a single shape across SDK
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/lib/selection.ts]: The payload shapes described here don't match the SDK: selection arrives via `annotationSelection.change` as an `Immutable.List` of annotations, and `viewState.change` carries a `ViewState` — never `{ annotation }`. The defensive normalization is fine to keep, but the comment describes shapes the SDK doesn't emit and will mislead anyone copying this helper.

File: web/viewer/web-sdk-demo/src/form-creator/index.ts [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+    const formField = await getFormFieldForAnnotation(instance, annotation)
+    if (!formField) {
+      console.warn('No backing form field found; skipping widget-only delete.')
+      return
+    }
+    const target = formField.id ?? formField
+    if (!target) return
+    await instance.delete(target)
+  } catch (error) {
+    console.error('Failed to delete field', error)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/form-creator/index.ts:244]: `deleteField`/`renameField` swallow SDK failures to `console.error` with no user feedback — the user clicks delete/rename, it silently fails, document unchanged. The rest of the demo surfaces failures (`Toolbar`'s `operationError` strip, `App`'s `window.alert`); route these through one of those. (The delete itself is correct — deleting the form field cascades to its widget annotations.)

File: web/viewer/web-sdk-demo/src/form-creator/index.ts [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+
+    if (!formField?.set || !annotation.set) {
+      console.warn('No backing form field found; skipping partial field rename.')
+      return
+    }
+
+    updates.push(formField.set('name', nextName))
+    updates.push(annotation.set('formFieldName', nextName))
+
+    if (updates.length > 0) await instance.update(updates)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/form-creator/index.ts]: The `updates` accumulator is dead: the early return guarantees both `.set` calls run, so two items are always pushed and `updates.length > 0` is never false.

```suggestion
    const formField = await getFormFieldForAnnotation(instance, annotation)

    if (!formField?.set || !annotation.set) {
      console.warn('No backing form field found; skipping partial field rename.')
      return
    }

    await instance.update([
      formField.set('name', nextName),
      annotation.set('formFieldName', nextName),
    ])
```

File: web/viewer/web-sdk-demo/src/signing/storage.ts [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+    const raw = localStorage.getItem(key)
+    if (!raw) return fallback
+    return JSON.parse(raw) as T
+  } catch {
+    return fallback
+  }
+}
+
+function writeJSON(key: string, value: unknown) {
+  localStorage.setItem(key, JSON.stringify(value))
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/signing/storage.ts]: `writeJSON` is unguarded while `readJSON` catches. Signature data-URLs are large base64 PNGs, so `setItem` realistically throws `QuotaExceededError` (or `SecurityError` when storage is disabled) — and since `saveSignature` runs inside a synchronous click handler with no error boundary, the throw escapes uncaught and the insert silently aborts. Wrap it like `readJSON` and surface a “couldn't save (storage full)” message.

File: web/viewer/web-sdk-demo/src/signing/SigningModal.tsx [hunk 23]
```diff
@@ -0,0 +1,403 @@
...
+    const file = e.target.files?.[0]
+    e.target.value = ''
+    if (!file) return
+    if (file.size > 1024 * 1024) {
+      alert('File must be smaller than 1MB.')
+      return
+    }
+    const reader = new FileReader()
+    reader.onload = () => onChange(typeof reader.result === 'string' ? reader.result : null)
+    reader.readAsDataURL(file)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/signing/SigningModal.tsx:326]: No `reader.onerror` — if the read fails (unreadable/corrupt file, IO error) `onChange` never fires, the Insert button stays disabled, and the user gets no feedback (the size-limit branch just above does `alert`). Add an `onerror` that clears state and surfaces the failure.

File: web/viewer/web-sdk-demo/src/ink/InkToolbar.tsx [hunk 24]
```diff
@@ -0,0 +1,392 @@
...
+    return () => {
+      instance.setViewState((vs) => vs.set('interactionMode', null))
+      instance.setCurrentAnnotationPreset(null)
+    }
+  }, [instance])
+
+  const undo = useCallback(async () => {
+    if (!instance) return
+    try {
+      await instance.history.undo()
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/ink/InkToolbar.tsx:133]: This `undo` (and the `deleteAll` handler) catch to `console.warn` only — the user clicks Undo/Delete, it fails, nothing happens, no indication why. `TextToolbar` mirrors this. Reuse the `operationError` pattern from `Toolbar` so these mutating actions report failure.

File: web/viewer/web-sdk-demo/src/App.tsx [hunk 25]
```diff
@@ -0,0 +1,430 @@
...
+          window.alert(
+            'Could not create this form field because the SDK form-field constructor is unavailable.',
+          )
+          return
+        }
+
+        await instance.create([widget, formField])
+        instance.setSelectedAnnotations?.(annotationIdList([widget.id]))
+      } catch (err) {
+        console.error('Could not place field', err)
```
Comment (ritz078) [inline web/viewer/web-sdk-demo/src/App.tsx:168]: This `create()` failure is `console.error`-only, but sibling paths in this same function `window.alert` on failure — so a field dropped off-page (or a license/validation rejection) just silently doesn't appear. Make the feedback consistent.

File: document-authoring/ai-editing/src/components/document-authoring-ai-shell.tsx [hunk 26]
```diff
@@ -836,7 +836,7 @@ export function DocumentAuthoringAiShell() {
               href="https://nutrient.io/sdk/document-authoring/"
               className="app-header-btn app-header-btn-outline"
             >
-              Learn More
+              Document Authoring SDK
```
Comment (ritz078) [inline document-authoring/ai-editing/src/components/document-authoring-ai-shell.tsx]: Is this change (plus the SVG `<title>` additions under `document-authoring/ai-editing/`) intentionally part of this PR? It's unrelated to the web-sdk-demo and isn't mentioned in the description — looks like it may have been swept in by a repo-wide format/lint run. Consider splitting it out.

**Review by ritz078 — APPROVED** [review]
Two items with no single line to anchor to:

- **`web/viewer/README.md` not updated.** `AGENTS.md` (Runnable Local Examples) requires updating the nearest category README so a new example is discoverable; the catalog currently lists only `multi-tab`.
- **No CI builds or typechecks this example** — `typecheck.yml` is scoped to `playground/**` and the demo sits in `biome.json`'s ignore list, so `tsc -b && vite build` isn't exercised in CI. Matches repo norms (multi-tab is the same); just verify the build locally before merge.
